src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c

```diff
@@ -511,13 +529,23 @@ static void rr_connectivity_changed_locked(grpc_exec_ctx *exec_ctx, void *arg,
                                            grpc_error *error) {
   subchannel_data *sd = arg;
   round_robin_lb_policy *p = sd->subchannel_list->policy;
+  if (GRPC_TRACER_ON(grpc_lb_round_robin_trace)) {
+    gpr_log(
+        GPR_DEBUG,
+        "[RR %p] connectivity changed for subchannel %p, subchannel_list %p: "
+        "prev_state=%d new_state=%d p->shutdown=%d "
+        "sd->subchannel_list->shutting_down=%d error=%s",
+        (void *)p, (void *)sd->subchannel_list, (void *)sd->subchannel,
+        sd->prev_connectivity_state, sd->curr_connectivity_state, p->shutdown,
+        sd->subchannel_list->shutting_down, grpc_error_string(error));
+  }
   // If the policy is shutting down, unref and return.
   if (p->shutdown) {
     rr_subchannel_list_unref(exec_ctx, sd->subchannel_list, "pol_shutdown");
     GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &p->base, "pol_shutdown");
     return;
   }
-  if (sd->subchannel_list->shutting_down) {
+  if (sd->subchannel_list->shutting_down && error == GRPC_ERROR_CANCELLED) {
```

Comment:
Yes. Again, concurrent updates. Consider:

1. Subchannel list X=[a,b,c] is created as part of an update.
2. Subchannels a, b, c have state INIT.
3. An update comes, subchannel list Y=[a, x, y]. Outdated subchannel list X is shutdown, we unsubscribe from connectivity updates for subchannels a, b, c.
4. rr_connectivity callback is called for "a". However, "a" is referenced by two separate `sd`s, one corresponding to X, another one to Y. In the context of the outdated list X, we only want to go inside this `if` when the rr_connectivity callback corresponds to the subchannel unsubscription from step 3. That is indicated by `error` being CANCELLED. The connectivity callback will be invoked for the "a" in Y with an `error == NONE` as it transitions from INIT to CONNECTING, for example.